apollo-federation/src/operation/optimize.rs

```diff
@@ -1549,6 +1552,16 @@ impl Operation {
         self.reuse_fragments_inner(fragments, Self::DEFAULT_MIN_USAGES_TO_OPTIMIZE)
     }
 
+    pub(crate) fn generate_fragments(&mut self) -> Result<(), FederationError> {
+        let mut generator = FragmentGenerator::default();
+        generator.visit_selection_set(&mut self.selection_set)?;
+        let fragments = generator.into_inner();
+        // To remove single-use fragment definitions:
+        // self.selection_set = fragments.reduce(&self.selection_set, 2)?;
```

Comment:
Do you plan to uncomment this line?

Reply:
I've removed it. This is something that we could do, but it doesn't make a big difference in query size, and I think a great benefit of fragment generation is how simple and obvious it is (including being a single pass, while reducing fragments requires looping until getting to a stable result).

We should consider it if/when we are doing more complicated extractions, where the entire fragment definition is overhead